tpmutil/mssim/mssim.go

```diff
@@ -0,0 +1,227 @@
+	}
+
+	return &Conn{addr: cmdAddr, dial: dial}, nil
+}
+
+// sendPlatformCommand sends device management commands to the simulator.
+//
+// See: "D.4.3.2. PlatformServer()"
+func sendPlatformCommand(conn net.Conn, buff [4]byte, u uint32) error {
+	binary.BigEndian.PutUint32(buff[:], u)
+	if _, err := conn.Write(buff[:]); err != nil {
+		return fmt.Errorf("write platform command: %v", err)
+	}
+	if _, err := io.ReadFull(conn, buff[:]); err != nil {
+		return fmt.Errorf("read platform command: %v", err)
+	}
+	rc := binary.BigEndian.Uint32(buff[:])
+	if rc != 0 {
+		return fmt.Errorf("unexpected platform command response code: 0x%x", rc)
+	}
+	return nil
+}
+
+const (
+	connStateIdle    = 0
+	connStateRunning = 1
+)
+
+// Conn is a Microsoft Simulator client that can be used as a connection for the
+// tpm2 package.
+//
+//		client, err := mssim.Open(mssim.Config{})
+//		if err != nil {
+//			// handle error
+//		}
+//		defer client.Close()
+//
+//		randBytes, err := tpm2.GetRandom(client, 8)
+//
+type Conn struct {
+	addr string
+
+	dial func(ctx context.Context, network, address string) (net.Conn, error)
+
+	// Track if there's an outgoing connection to detect concurrent use.
+	connState uint32
+	// Cached connection
+	conn net.Conn
+}
+
+// Read is implemented so Conn can be passed as an io.ReadWriter, but is
+// unimplemented and will always return an error. See RunRawCommand to read
+// responses from the simulator.
+func (t *Conn) Read(b []byte) (int, error) {
+	return 0, errors.New("mssim transport: read not supported")
+}
+
+// Write is implemented so Conn can be passed as an io.ReadWriter, but is
+// unimplemented and will always return an error. See RunRawCommand to write
+// responses to the simulator.
+func (t *Conn) Write(b []byte) (int, error) {
+	return 0, errors.New("mssim transport: write not supported")
+}
+
+// RunRawCommand executes a serialized TPM2 command and returns the serialized
+// response.
+func (t *Conn) RunRawCommand(b []byte) ([]byte, error) {
+	// RunRawCommand isn't safe for concurrent calls. Be extra defensive and
+	// panic if it's called by multiple goroutines.
+	swapped := atomic.CompareAndSwapUint32(&t.connState, connStateIdle, connStateRunning)
+	if !swapped {
+		panic("concurrent calls to RunRawCommand")
+	}
+	defer func() { atomic.StoreUint32(&t.connState, connStateIdle) }()
+
+	// Check for cached connection.
+	conn := t.conn
+	if conn == nil {
+		c, err := t.dial(context.TODO(), "tcp", t.addr)
+		if err != nil {
+			return nil, fmt.Errorf("dial command endpoint: %v", err)
+		}
+		conn = c
+	}
+
+	resp, err := runRawCommand(conn, b)
+	if err == nil {
+		t.conn = conn
+	} else {
+		t.conn = nil
+		conn.Close()
+	}
+	return resp, err
+}
+
+// runRawCommand frames a TPM2 command and executes it against the simulator
+// returning the unwrapped framed response.
+//
+// See: D.4.3.12. TpmServer()
+func runRawCommand(conn io.ReadWriter, b []byte) ([]byte, error) {
+	req := make([]byte, 4+1+4+len(b))
```

Comment:
Use `bytes.Buffer` and `binary.Write` instead of hardcoding lengths

Reply:
done